pkg/ssh/manager.go

```diff
@@ -204,43 +235,44 @@ func (sm *StreamManager) updateRecordsBinding(ctx context.Context, _ uint64, rec
 		if record.DeletedAt == nil {
 			continue
 		}
-		// a session was deleted; terminate all of its associated streams
-		streams := sm.bindingStreams[record.Id]
-		for streamID := range streams {
+		// Session binding was deleted; terminate all of its associated streams
+		for streamID := range sm.bindingStreams[record.Id] {
 			log.Ctx(ctx).Debug().
 				Str("session-id", record.Id).
 				Uint64("stream-id", streamID).
-				Msg("terminating stream: session revoked")
+				Msg("terminating stream: session binding revoked")
 			sm.terminateStreamLocked(streamID)
 		}
 		delete(sm.bindingStreams, record.Id)
 	}
 }
 
-func (sm *StreamManager) SetSessionIDForStream(ctx context.Context, streamID uint64, sessionID string, sessionBindingID string) error {
-	sm.mu.Lock()
-	for !sm.initialSyncDone {
-		sm.mu.Unlock()
-		select {
-		case <-sm.waitForInitialSync:
-		case <-time.After(10 * time.Second):
-			return errors.New("timed out waiting for initial sync")
-		case <-ctx.Done():
-			return context.Cause(ctx)
-		}
-		sm.mu.Lock()
+func (sm *StreamManager) OnStreamAuthenticated(ctx context.Context, streamID uint64, req AuthRequest) error {
+	if err := sm.waitForInitialSync(ctx); err != nil {
+		return err
 	}
-	defer sm.mu.Unlock()
-	if sm.sessionStreams[sessionID] == nil {
-		sm.sessionStreams[sessionID] = map[uint64]struct{}{}
+	activeStream := sm.activeStreams[streamID]
+	if activeStream.Session != nil || activeStream.SessionBindingID != nil {
+		return status.Errorf(codes.Internal, "stream %d already has an associated session", streamID)
+	}
+
+	if sm.sessionStreams[req.SessionID] == nil {
+		sm.sessionStreams[req.SessionID] = map[uint64]struct{}{}
 	}
-	if sm.bindingStreams[sessionBindingID] == nil {
-		sm.bindingStreams[sessionBindingID] = map[uint64]struct{}{}
+	if sm.bindingStreams[req.SessionBindingID] == nil {
+		sm.bindingStreams[req.SessionBindingID] = map[uint64]struct{}{}
 	}
-	sm.sessionStreams[sessionID][streamID] = struct{}{}
-	sm.bindingStreams[sessionBindingID][streamID] = struct{}{}
-	sm.activeStreams[streamID].Session = &sessionID
-	sm.activeStreams[streamID].SessionBindingID = &sessionBindingID
+	sm.sessionStreams[req.SessionID][streamID] = struct{}{}
+	sm.bindingStreams[req.SessionBindingID][streamID] = struct{}{}
+
+	activeStream.Session = new(string)
+	*activeStream.Session = req.SessionID
+	activeStream.SessionBindingID = new(string)
+	*activeStream.SessionBindingID = req.SessionBindingID
+
+	activeStream.PortForwardManager.AddUpdateListener(activeStream.Handler)
```

Comment:
we hold lock ni `sm.waitForInitialSync` but we do not lock here, why?

Reply:
Good catch, this is missing a lock.